File: labs/29-deployment/python/main.py
import os
import sys
from pathlib import Path

from fastapi import FastAPI, HTTPException
from pydantic import BaseModel

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "shared" / "config"))

from openai_client import OpenAIClient, configure_logging

configure_logging()

app = FastAPI(title="Agent Lab 29: Deployment")

# Validate config at startup and fail gracefully.
api_key = os.getenv("OPENAI_API_KEY", "")
if not api_key:
    print("Warning: OPENAI_API_KEY is not set. /chat will return 503 until configured.", file=sys.stderr)

client = OpenAIClient()


class ChatRequest(BaseModel):
    message: str
    model: str | None = None


class ChatResponse(BaseModel):
    reply: str
    model: str


@app.get("/health")
def health() -> dict:
    return {"status": "ok", "configured": bool(api_key)}


@app.post("/chat", response_model=ChatResponse)
def chat(request: ChatRequest) -> ChatResponse:
    if not api_key:
        raise HTTPException(status_code=503, detail="OPENAI_API_KEY is not configured")

    if not request.message.strip():
        raise HTTPException(status_code=400, detail="message is required")

    try:
        response = client.chat_completion(
            messages=[{"role": "user", "content": request.message}],
            model=request.model,
            max_tokens=300,
            temperature=0.0,
        )
        reply = response["choices"][0]["message"].get("content", "")
        return ChatResponse(reply=reply, model=request.model or client.model)
    except Exception as exc:
        raise HTTPException(status_code=502, detail=f"Upstream error: {exc}") from exc


def main():
    import uvicorn

    uvicorn.run("main:app", host="0.0.0.0", port=8000, reload=False)


if __name__ == "__main__":
    main()

File: labs/29-deployment/java/Main.java
import com.agentlabs.OpenAiClient;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;

import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.util.Map;

public class Main {
    private static final ObjectMapper mapper = new ObjectMapper();
    private static String apiKey;

    public static void main(String[] args) throws Exception {
        apiKey = System.getenv().getOrDefault("OPENAI_API_KEY", "");
        if (apiKey.isEmpty()) {
            System.err.println("Warning: OPENAI_API_KEY is not set. /chat will return 503 until configured.");
        }

        OpenAiClient client = new OpenAiClient();
        int port = Integer.parseInt(System.getenv().getOrDefault("PORT", "8000"));
        HttpServer server = HttpServer.create(new InetSocketAddress(port), 0);

        server.createContext("/health", exchange -> {
            sendJson(exchange, 200, mapper.writeValueAsString(Map.of("status", "ok", "configured", !apiKey.isEmpty())));
        });

        server.createContext("/chat", exchange -> {
            if (!"POST".equalsIgnoreCase(exchange.getRequestMethod())) {
                sendJson(exchange, 405, mapper.writeValueAsString(Map.of("error", "Method not allowed")));
                return;
            }
            if (apiKey.isEmpty()) {
                sendJson(exchange, 503, mapper.writeValueAsString(Map.of("error", "OPENAI_API_KEY is not configured")));
                return;
            }
            JsonNode req = mapper.readTree(exchange.getRequestBody());
            String message = req.has("message") ? req.get("message").asText() : "";
            String model = req.has("model") ? req.get("model").asText() : "";
            if (message.isBlank()) {
                sendJson(exchange, 400, mapper.writeValueAsString(Map.of("error", "message is required")));
                return;
            }
            try {
                ObjectNode payload = client.buildMessages(message);
                if (!model.isEmpty()) payload.put("model", model);
                payload.put("max_tokens", 300);
                payload.put("temperature", 0.0);
                ObjectNode response = client.chatCompletion(payload);
                String reply = client.extractMessage(response).get("content").asText();
                sendJson(exchange, 200, mapper.writeValueAsString(Map.of("reply", reply, "model", model.isEmpty() ? "gpt-4o-mini" : model)));
            } catch (Exception e) {
                sendJson(exchange, 502, mapper.writeValueAsString(Map.of("error", "Upstream error: " + e.getMessage())));
            }
        });

        server.start();
        System.out.println("Server listening on :" + port);
    }

    private static void sendJson(HttpExchange exchange, int status, String body) throws IOException {
        exchange.getResponseHeaders().set("Content-Type", "application/json");
        byte[] bytes = body.getBytes();
        exchange.sendResponseHeaders(status, bytes.length);
        try (OutputStream os = exchange.getResponseBody()) {
            os.write(bytes);
        }
    }
}
